Add documentation for HA Kubernetes

brendandburns opened the pull request:
@lavalamp @davidopp @smarterclayton 

I plan on adding some pretty pictures before submitting, but here for some initial review.

brendandburns committed:
Add documentation for HA Kubernetes

brendandburns commented:
Depends on #9133

bgrant0607 commented:
FYI: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/proposals/high-availability.md

smarterclayton commented:
@derekwaynecarr fyi

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,126 @@
+
+## Introduction
+This document describes how to build a high-available Kubernetes cluster.  This is a fairly advanced topic, generally useful to
+people who need to set up a higly availble, redundant Kubernetes cluster for production workloads.  Users who merely want to
```
I think setting this up as a dichotomy of HA vs. {GKE, single-node local} is a little confusing. I would just replace this sentence with "Users who merely want to experiment with Kubernetes are encouraged to use a non-HA configuration. See [URL] for a list of getting started guides."

brendandburns replied:
Addressed, modified using reliable storage as "extra redundancy" rather than the default.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,126 @@
+   * Setting up a redundant, reliable storage layer
+   * Installing a health-checking daemon runner
+   * Turning up a clustered instance of the etcd key/value store
+   * Starting a replicated, load balanced kubernetes API servers
```
remove "a"

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,126 @@
+   * Installing a health-checking daemon runner
+   * Turning up a clustered instance of the etcd key/value store
+   * Starting a replicated, load balanced kubernetes API servers
+   * Setting up master-elected kubernetes components
```
"components" is too vague -- spell them out (controller manager, scheduler)

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,126 @@
+The remainder of this guide assumes that you are setting up a 3-node clustered master, where each machine is running some flavor of Linux.
+Examples in the guide are given for Debian distributions, but they should be easily adaptable to other distributions.
+
+## Reliable storage
```
I don't fully agree with this section. One shouldn't need reliable storage if one is using clustered etcd and the probability of all etcd replicas losing their data is acceptably small. On bare metal this would mean all the disks breaking or being erased, and on cloud this would mean all the VMs losing their per-VM data.

brendandburns replied:
Addressed, modified using reliable storage as "extra redundancy" rather than the default.

davidopp commented on docs/high-availability.md:
````diff
@@ -0,0 +1,126 @@
+choices.  We also need to health check and restart the docker daemon, we'll use ```monit``` for this as well.
+
+You should install the [kubelet binary](https://storage.googleapis.com/kubernetes-release/release/v0.18.0/bin/linux/amd64/kubelet), the
+[/etc/init.d/kubelet](high/availability/init-kubelet) and [/etc/default/kubelet](high-availability/default-kubelet)
````
high/availability -> high-availability

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,126 @@
+
+For simplicity, we'll use etcd's built in discovery to build our cluster.
+
+First, hit the etcd discovery service to create a new toke:
```
toke -> token

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
````diff
@@ -0,0 +1,126 @@
+for ```${NODE_IP}``` on each machine.
+
+Once you copy this into all three nodes, you should have a clustered etcd set up.  You can validate with
+```etcdctl member list``` and ```etcdctl cluster-health```.
````
Put the text on separate lines from the three backticks, otherwise it confuses the Github PR review colorizer.

brendandburns replied:
done.

davidopp commented:
Something you could make clearer in several spots is mentioning explicitly which steps trigger binaries to start vs. when you need to start stuff up manually.

IIUC all of the steps that copy stuff into /etc/kubernetes/manifests/ start up something -- you should say that. (For example, 
```
On each node, copy the [etcd.manifest](high-availability/etcd.manifest) file into /etc/kubernetes/manifests/etcd.manifest
Kubelet will start up etcd on a node when it notices the new manifest file.
```
)
and likewise for API server and modmaster.

Likewise, the instructions at the end of "Reliable nodes" don't distinguish steps that just copy bits vs. steps that start up processes. I think you could make those instructions a bit more step-by-step and clearer.

davidopp commented:
(You say things like "At this point you should have XXX" but it's ambiguous what the user did that caused XXX to start.)

brendandburns commented:
Tried to address the comments about making it explicit what the kubelet starts.

Addressed other comments.

I plan on adding an image (or two) this evening as well, will ping the PR when that is landed.

brendandburns commented:
Ok, I updated the doc to add an overview PNG (and added the SVG for good measure too ;)

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+# High Availability Kubernetes Clusters
+
+## Introduction
+This document describes how to build a high-available Kubernetes cluster.  This is a fairly advanced topic, 
```
, at end of line should be .

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+
+## Introduction
+This document describes how to build a high-available Kubernetes cluster.  This is a fairly advanced topic, 
+Users who merely want to experiment with Kubernetes are encouraged to use a configurations that are simpler to set up such as
```
a configurations -> configurations

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+or try [Google Container Engine](https://cloud.google.com/container-engine/) for hosted Kubernetes.
+
+## Overview
+Setting up a truly reliable, highly available distributed system requires a number of distinct steps, it is akin to
```
I'm not sure "distinct" adds anything to the sentence.

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+
+## Initial set-up.
+The remainder of this guide assumes that you are setting up a 3-node clustered master, where each machine is running some flavor of Linux.
+Examples in the guide are given for Debian distributions, but they should be easily adaptable to other distributions.
```
Also mention that this should work on bare metal or cloud?

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+The remainder of this guide assumes that you are setting up a 3-node clustered master, where each machine is running some flavor of Linux.
+Examples in the guide are given for Debian distributions, but they should be easily adaptable to other distributions.
+
+The easiest way to implement an HA Kubernetes cluster, is to start with an existing single-master cluster.  See [https://get.k8s.io](https://get.k8s.io)
```
don't need a comma after "cluster"
Also, make it clearer you actually want them to run the script at get.k8s.io. Suggested fix: say "Follow the instructions at ..." rather than "See ..."

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+to protect the data.  Whatever else happens, whatever catches on fire, if you have the data, you can rebuild.  If you lose the data, you're
+done.
+
+Clustered etcd, already replicates your storage to three nodes in your cluster.  This means that to lose data, all three nodes would need
```
no need for comma after etcd
also, there are some assumptions baked into "Clustered etcd, already replicates your storage to three nodes in your cluster" (like, that you set up three nodes, told etcd to run on those three nodes, etc.) that aren't obvious at this point in the text. I'd suggest instead saying "Clustered etcd already replicates your storage to all of its instances. This means that to lose data, all etcd nodes would need..."

brendandburns replied:
done.

timothysc replied:
+1, nix the 3 node bit.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+
+Clustered etcd, already replicates your storage to three nodes in your cluster.  This means that to lose data, all three nodes would need
+to have their physical (or virtual) disks fail at the same time.  The probability that this occurs is relatively low, so for many people
+running a replicated etcd cluster, with periodic off-site backups, of course, is likely reliable enough.
```
I would remove "off-site." I think off-site backup is useful when you want to save a lot of history. But we're talking about cluster state here, which really isn't useful once it gets more than a few hours out-of-date (at most). Once you get older than that, you really don't have any choice but to wipe the cluster and start from empty. In Borg we have never restored from anything other than the most recent checkpoint available that wasn't on the Borgmaster's disks (and this is an every-five-years kind of event); this was always on spinning disk somewhere, not on the masters (never tape).

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
````diff
@@ -0,0 +1,177 @@
+for each node.  Throughout these instructions, we assume that this storage is mounted to your machine in ```/var/etcd/data```
+
+## Reliable nodes
+On each master, we are going to run a number of processes that implement the Kubernetes API.  The first step in making these reliable is
````
master -> master node

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+
+## Reliable nodes
+On each master, we are going to run a number of processes that implement the Kubernetes API.  The first step in making these reliable is
+to make sure that each automatically restarts when they fail.  To achieve this, we need to install a process watcher.  We choose to use
```
when it fails

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+[/etc/monit/conf.d/docker](high-availability/monit-docker) configs.
+
+## Clustered etcd
+Ok, you now have reliable storage, and reliable nodes, now it's time to start building a cluster.
```
Nit: since the user hasn't started etcd yet, they don't have reliable storage yet.

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
````diff
@@ -0,0 +1,177 @@
+The kubelet on each node actively monitors the contents of that directory, and it will create an instance of the ```etcd```
+server from the definition of the pod specified in ```etcd.manifest```.
+
+Note that you should substitute the token URL you got above for ```${DISCOVERY_TOKEN}``` on all three machines,
````
Note that you should -> Note that in etcd.manifest you should

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+### Load balancing.
+At this point, you should have 3 apiservers all working correctly.  If you set up a network load balancer, you should
+be able to access your cluster via that load balancer, and see traffic balancing between the apiserver instances.  Setting
+up a load balancer will depend on the specifics of your load balancer, for example instructions for the Google Cloud
```
of your load balancer -> of your platform

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+
+## Master-elected components
+So far we have set up state storage, and we have set up the API server, but we haven't run anything that actually modifies
+cluster state.  To achieve this reliably, we only want to have one actor modifying state at a time, but we want replicated
```
cluster state -> cluster state, such as the controller manager and scheduler.

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+So far we have set up state storage, and we have set up the API server, but we haven't run anything that actually modifies
+cluster state.  To achieve this reliably, we only want to have one actor modifying state at a time, but we want replicated
+instances of these actors, in case a machine dies.  To achieve this, we are going to use a lease-lock in etcd to perform
+master election.  We'll use a simple utility that runs the master election, and then either creates or deletes a pod
```
Maybe mention that in the future, we will use something more tightly integrated, and point to 
 https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/proposals/high-availability.md ?

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
````diff
@@ -0,0 +1,177 @@
+### Running the podmaster
+Now that the configuration files are in place, copy the [podmaster.manifest](high-availability/podmaster.manifest) config file into ```/etc/kubernetes/manifests/```
+
+As before, the kubelet on the node monitors this directory, and will start an instance of the podmaster using the pod specification provided in ```podmaster.manifest```.
````
Maybe redundant, but I would end this section with something like:

Now you will have one scheduler process running and one controller manage process running, and if one of these processes or its nodes fails, another node will launch the corresponding process to take over.

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+If you have an existing cluster, this is as simple as reconfiguring your kubelets to talk to the load-balanced endpoint, and
+restarting the kubelets on each node.
+
+If you are turning up a fresh cluster, you will need to install the kubelet and kube-proxy on each worked node, and
```
worker node

brendandburns replied:
done.

davidopp commented:
You might want to add docker to the picture of the master node, for completeness, but I don't think it's necessary.

bgrant0607 commented:
cc @timothysc

timothysc commented:
cc @rrati

mattf commented on docs/high-availability.md:
```diff
@@ -0,0 +1,177 @@
+Note, if you are using authentication, you may need to regenerate your certificat to include the IP address of the balancer,
+in addition to the IP addresses of the individual nodes.
+
+## Master-elected components
```
this is a really slick approach. using a kubelet managed pod that manages the active scheduler/controller-manager by adjusting kubelet config and coordinating through etcd.

due to delays, you can have a situation where one kubelet is about to kill the active component while another kubelet has already started up the new active component. in this case you have two active components. what guarantees are in place to avoid state corruption or incorrect behavior (e.g. over/under provisioning w/ repl controllers, or double pod assignment by scheds)?

brendandburns replied:
Practically, it's unlikely that this will occur, since the lease is re-acquired @ 1/2 TTL, practically this could be caused by either:
   * A super flaky network, such that lease acquisitions fail, but controller/scheduler actions succeed
   * A massively overloaded machine, such that the pod master never gets any cycles to run.

Of course, in any real world environment, if it can happen, it _will_ happen, so in that vein:

## Scheduler
The scheduler won't double schedule, since the scheduler will only schedule pods that have an empty ```Host``` field, and we use compare-and-swap with a resource version to ensure that we get atomic updates of that ```Host``` field.  This prevents read-update-write races.  It is also possible that you could over-schedule a specific node, but in that case the kubelet throws the Pod into the ```Failure``` state, and your replication controller should schedule a new instance of the pod, since it sees desired state doesn't match current state.

## Controller Manager
The controller-manager is a little bit trickier, with multiple controller managers running, you may see replication controllers fighting with each other, e.g. Desired # is 3, current is 2, both controllers add one.  Now current is 4, both controllers delete, ...

Given that we expect the a double controller-manager situation to eventually resolve itself, this pattern will only continue for a brief period of time.  However, solving this issue is one of the major reasons for driving the lease-locking deeper into the controller-manager binary.

(we could also achieve the same thing with compare-and-swap, once etcd supports cross-object preconditions)

mattf replied:
thanks for the explanation.

for a pod w/o a replication controller, you can get a Failure for no fault of the user. however, a user not using a repl ctrlr should be handling potential random failures anyway.

are there other cases i may have not listed?

mattf commented on docs/high-availability.md:
````diff
@@ -0,0 +1,177 @@
+The kubelet monitors this directory, and will automatically create an instance of the ```kube-apiserver``` container using the pod definition specified
+in the file.
+
+### Load balancing.
````
for load balancing, would you suggest kubelets, cli, scheduler, controller-manager are configured w/ the load-balancer endpoint?

for internal clients (introspective services), this section should explain configuration needed after https://github.com/GoogleCloudPlatform/kubernetes/issues/6572

brendandburns replied:
Added some further text to describe this.

davidopp replied:
extraneous period after "balancing" in the title

brendandburns commented:
Comments addressed, please re-check.

Thanks!
--brendan

erictune commented on docs/high-availability.md:
```diff
@@ -0,0 +1,193 @@
+Users who merely want to experiment with Kubernetes are encouraged to use configurations that are simpler to set up such as
+the simple [Docker based single node cluster instructions](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/getting-started-guides/docker.md),
+or try [Google Container Engine](https://cloud.google.com/container-engine/) for hosted Kubernetes.
+
```
Not only are they simpler, but they are continuously tested.  I'm sure we'll get to the point where the HA is part of e2e, but until then, best to make it clear which path is best supported.

brendandburns replied:
added a note to make this point.

timothysc commented:
cc @eparis - implies deployment of master components as containers.

brendandburns commented:
@timothysc this is already the way they are deployed @ head [at least for GCE, GKE]

timothysc commented:
@brendandburns - we've been meaning to make the transition on Atomic, this helps to force the change ;-).

mattf commented on docs/high-availability.md:
````diff
@@ -0,0 +1,196 @@
+
+For pods that you deploy into the cluster, the ```kubernetes``` service/dns name should provide a load balanced endpoint for the master automatically.
+
+For external users of the API (e.g. the ```kubectl``` command line interface, continuous build pipelines, or other clients) you will want to configure
````
will you add the scheduler and controller-manager are examples of other client?

brendandburns replied:
actually, scheduler and controller-manager just connect to localhost.  There is no world in which they can be successfully running on a master node replica, and there isn't a corresponding api server.

eventually we may self-host those containers in the k8s cluster itself, in which case we'd use the load balanced endpoing, but we're not there yet.

mattf replied:
the sched&cm have a --master arg to talk to a non-localhost apiserver, but it makes sense that it's preferred to run against the local apiserver. i ran into this in #7181.

mattf commented on docs/high-availability.md:
```diff
@@ -0,0 +1,196 @@
+
+## Overview
+Setting up a truly reliable, highly available distributed system requires a number of steps, it is akin to
+wearing underwear, pants, a belt, suspenders, another pair of underwear, and another pair of pants.  We go into each
```
this is amusing. i'd say it's more like keeping a spare outfit in your backpack and not having anywhere discrete to change!

brendandburns commented:
shippable is a flake, fwiw.

mattf commented on docs/high-availability.md:
```diff
@@ -0,0 +1,196 @@
+instructions at [https://get.k8s.io](https://get.k8s.io)
+describe easy installation for single-master clusters on a variety of platforms.
+
+## Reliable storage
```
this section could be simplified to roughly state: 0. kubernetes relies on etcd for reliable storage, make sure your comfortable w/ your etcd deployment; 1. though etcd upstream touches on this, you can use reliable volumes attached to your nodes (like GPD, Gluster, Ceph, etc) or ISCSI, RAID, etc.; 2. a pitfall you should avoid is...[putting etcd data in /tmp?]

but 1. and 2. could be optional.

also, the "clusterd etcd" section could be moved up and merged here.

imho, the story is "kube needs etcd" "make etcd ha (here's a link)" "if you want extra assurances use non-local reliable or RAIDed storage"

davidopp replied:
Yeah, I think it's confusing that you split the discussion of etcd between this section and the etcd section. It made sense in the first version of the doc, where this section was telling you that you had to use PD or iSCSI/etc. and didn't really talk about etcd. But now it's confusing. I would suggest to merge the two sections; whether the merged section goes before or after "reliable nodes" isn't particularly important.

brendandburns replied:
Tried to fix this up...  See what you think.

davidopp replied:
Did you push new commit?

mattf commented on docs/high-availability.md:
````diff
@@ -0,0 +1,196 @@
+You should install the [kubelet binary](https://storage.googleapis.com/kubernetes-release/release/v0.18.0/bin/linux/amd64/kubelet), the
+[/etc/init.d/kubelet](high-availability/init-kubelet) and [/etc/default/kubelet](high-availability/default-kubelet)
+scripts.  You should also install the monit daemon (```apt-get install monit```) and the [/etc/monit/conf.d/kubelet](high-availability/monit-kubelet) and
+[/etc/monit/conf.d/docker](high-availability/monit-docker) configs.
````
suggestion -
"
...first step make these reliable...need a watcher process.

for debian, we choose monit....

for systemd-based systems, you can run 'systemctl enable kubelet'. you should also have run 'systemctl enable docker'
"

brendandburns replied:
done.

mattf commented:
general comment re sched/cm failures. for a non-byzantine fault, the kubelet will perform a restart and all will be right with the world. byzantine faults can be split in two: honest healthz and dishonest healthz. dishonest healthz should be rare and require operator intervention. honest healthz can be handled by a liveness monitor, but isn't currently covered here. this is all important, because the meaning of "active" master should be that all the components of the master are active.

another complication, worthy of note, is the interaction w/ everyone's favorite oom killer. it can always cause a failure, but there are protections that can be taken to divert its attention away from the kubelet/podmaster/monit.

brendandburns commented:
Ok, I believe I have addressed all comments and this is ready to merge.

I have added liveness checks to the scheduler and controller manager to help with some of the byzantine fault cases.  I also added a pro-active lease release (based on timestamps) to podmaster in a separate PR.  That should help with some of the network partition edge cases.

Everyone please take look, I'd love to see this get merged sometime this week.

davidopp commented on docs/high-availability.md:
````diff
@@ -0,0 +1,199 @@
+## Reliable nodes
+On each master node, we are going to run a number of processes that implement the Kubernetes API.  The first step in making these reliable is
+to make sure that each automatically restarts when it fails.  To achieve this, we need to install a process watcher.  We choose to use
+the ```kubelet``` that we use on each of the nodes.  This is convenient, since we can use containers to distribute our binaries, we can
````
The sentence "We choose to use the kubelet that we use on each of the nodes" is a great example of why we should bring back the "minion" term. (It's not clear which nodes you're talking about -- slave nodes, master nodes, or both.) I'm not suggesting a change here, just lobbying to bring back "minion" :-)

mattf replied:
nit, use the kubelet that is already running on each of the nodes

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+itself (insert who watches the watcher jokes here).  For Debian systems, we choose monit, but there are a number of alternate
+choices. For example, on systemd-based systems (e.g. RHEL, CentOS), you can run 'systemctl enable kubelet'. 
+
+You should install the [kubelet binary](https://storage.googleapis.com/kubernetes-release/release/v0.18.0/bin/linux/amd64/kubelet), the
```
Mention that this should already be running on the master node if they did a regular single-node setup?

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+
+### Clustering etcd
+The full details of clustering etcd are beyond the scope of this document, lots of details are given on the
+[etcd clustering page](https://github.com/coreos/etcd/blob/master/Documentation/clustering.md).  This example wals through
```
wals -> walks

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+on a different node.
+
+### Even more reliable storage
+Of course, if you are interested in increased data reliability, there are further options (belts *and* suspenders, ftw!).
```
maybe "there are further options, which make the place where etcd stores its data even more reliable than a regular disk (belts *and* ...)"

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+up a load balancer will depend on the specifics of your platform, for example instructions for the Google Cloud
+Platform can be found [here](https://cloud.google.com/compute/docs/load-balancing/)
+
+Note, if you are using authentication, you may need to regenerate your certificat to include the IP address of the balancer,
```
certificat -> certificate

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+cluster state, such as the controller manager and scheduler.  To achieve this reliably, we only want to have one actor modifying state at a time, but we want replicated
+instances of these actors, in case a machine dies.  To achieve this, we are going to use a lease-lock in etcd to perform
+master election.  We'll use a simple utility that runs the master election, and then either creates or deletes a pod
+depending on if it is the master or not.
```
the "it" is ambiguous -- maybe rephrase as something like "On each master node we'll run a simple utility that participates in the master election protocol, and then either creates or deletes a pod depending on whether that node is the master."

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+
+In the future, we expect to more tightly integrate this lease-locking into the scheduler and controller-manager binaries directly, as described in the [high availability design proposal](proposals/high-availability.md)
+
+### Installing configuration files.
```
extraneous period at end of heading

brendandburns replied:
done.

davidopp commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+node.
+
+## Conclusion
+At this point, you are done (yeah!) with the master components, but you still need to add nodes (boo!).
```
again an argument for "minions"...

brendandburns replied:
ack.

davidopp commented:
I made a few minor comments, but other than that LGTM.

mattf commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+# High Availability Kubernetes Clusters
+
+## Introduction
+This document describes how to build a high-available Kubernetes cluster.  This is a fairly advanced topic.
```
high-availability or highly-available

brendandburns replied:
done.

mattf commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+   * Setting up master-elected Kubernetes scheduler and controller-manager daemons
+
+Here's what the system should look like when it's finished:
+![High availability Kubernetes diagram](high-availability/ha.png)
```
nit, could add kubelet to the worker node boxes

mattf commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+
+Ready? Let's get started.
+
+## Initial set-up.
```
extraneous period

brendandburns replied:
removed (here and elsewhere)

mattf commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+Likewise, this set up should work whether you are running in a public or private cloud provider, or if you are running
+on bare metal.
+
+The easiest way to implement an HA Kubernetes cluster is to start with an existing single-master cluster.  The
```
nit, previously used high-availability / highly-available and switched to HA here

brendandburns replied:
added parenthetical "HA" to make this transition evident.

mattf commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+wearing underwear, pants, a belt, suspenders, another pair of underwear, and another pair of pants.  We go into each
+of these steps in detail, but a summary is given here to help guide and orient the user.
+
+The steps involved are as follows:
```
nit, this flows better if each of these steps clearly corresponds to a section below. you could use the same text or link them. currently "installing a health-checking daemon runner" maps to "reliable nodes" while "setting up a redundant, reliable..." maps to "establishing a redundent, ..."

brendandburns replied:
links added.

mattf commented on docs/high-availability.md:
```diff
@@ -0,0 +1,199 @@
+So far we have set up state storage, and we have set up the API server, but we haven't run anything that actually modifies
+cluster state, such as the controller manager and scheduler.  To achieve this reliably, we only want to have one actor modifying state at a time, but we want replicated
+instances of these actors, in case a machine dies.  To achieve this, we are going to use a lease-lock in etcd to perform
+master election.  We'll use a simple utility that runs the master election, and then either creates or deletes a pod
```
total aside - i always thought the word "election" was kinda bogus in this context, it's closer to sword in the stone type selection than an election

mattf commented on docs/high-availability.md:
````diff
@@ -0,0 +1,199 @@
+
+
+### Running the podmaster
+Now that the configuration files are in place, copy the [podmaster.manifest](high-availability/podmaster.manifest) config file into ```/etc/kubernetes/manifests/```
````
the podmaster.manifest appears to be missing from the commit

brendandburns replied:
it was mis-named.  fixed.

mattf commented on docs/high-availability.md:
````diff
@@ -0,0 +1,199 @@
+```
+
+Next, set up the descriptions of the scheduler and controller manager pods on each node.
+by copying [kube-scheduler.manifest](high-availability/kube-scheduler.manifest) and [kube-controller-manager.manifest](high-availability/kube-controller-manager.manifest) into ```/srv/kubernetes/...```
````
i assume the /srv/kubernetes path is referenced from the podmaster.manifest. should it have the ellipsis at the end?

brendandburns replied:
elipsis removed and sentence expanded to make it clear that you should copy it into the directory.

mattf commented:
looking good. feel free to ignore my nits.

mattf commented:
fyi, liveness checks are in #9383

mikedanese commented:
ref #473

davidopp commented:
@brendanburns should we merge this?

timothysc commented:
@rrati @mattf it would be nice to outline our pacemaker solution as an option.

jayunit100 commented:
Maybe we can merge this along with #9904  together.   The is a fully automated implementation of this doc, so can serve as a living breathing specification which we can `vagrant up` evolve over time

timothysc commented:
@jayunit100 
I think pull in the ref into this doc, but folks are definitely going to want to do their own thing.  e.g. - We did our own thing with pacemaker.

jayunit100 commented:
Yup 
- pacemaker impl should be added in also
- Best to merge this as first pass so folks can start extending.

jayunit100 commented:
hi reviewers: just a thought ~  can we concurrrently merge this with #9904 (the automated implementation for testing) ?

davidopp commented:
Brendan will be back from vacation on Monday; let's see what he wants to do.

brendandburns commented:
Comments addressed, ptal.

Thanks!
--brendan

davidopp commented:
LGTM

mikedanese commented on docs/high-availability/etcd.manifest:
```diff
@@ -0,0 +1,104 @@
+{
+"apiVersion": "v1beta3",
```
fyi v1beta3 is going to be dropped before v1 #10501

jayunit100 replied:
I think that's okay, we can update to v1 and merge this ;Then changes made will be valuable to people .

mattf commented:
lgtm

mikedanese commented:
LGTM

davidopp commented:
This just documentation and now has multiple LGTMs so marking ok-to-merge.
